File: NOVEMBER-2025/23-11-2025/Maximum Stone Removal.js
class Solution {
    maxRemove(stones) {
        const parent = new Map();

        const find = (x) => {
            if (!parent.has(x)) parent.set(x, x);
            if (parent.get(x) !== x) parent.set(x, find(parent.get(x)));
            return parent.get(x);
        };

        const union = (x, y) => {
            const px = find(x);
            const py = find(y);
            if (px !== py) parent.set(px, py);
        };

        for (const [x, y] of stones) {
            // prefix 'r' for row, 'c' for column to avoid collisions
            union(`r${x}`, `c${y}`);
        }

        const components = new Set();
        for (const [x, y] of stones) {
            components.add(find(`r${x}`));
        }

        return stones.length - components.size;
    }
}

File: NOVEMBER-2025/23-11-2025/Maximum Stone Removal.py
class Solution:
    def maxRemove(self, stones):
        parent = {}

        def find(x):
            if x not in parent:
                parent[x] = x
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            px, py = find(x), find(y)
            if px != py:
                parent[px] = py

        for x, y in stones:
            union(x, y + 10001)

        components = set()
        for x, y in stones:
            components.add(find(x))

        return len(stones) - len(components)